internal/state/responders.go

```diff
@@ -99,12 +112,93 @@ func (u *exists) handle(ctx context.Context, tx *ent.Tx, snap *snapshot) ([]resp
 	return res, nil
 }
 
-func (u *exists) getMessageID() imap.InternalMessageID {
-	return u.messageID
+func (u *targetedExists) getMessageID() imap.InternalMessageID {
+	return u.resp.messageID.InternalID
 }
 
-func (u *exists) String() string {
-	return fmt.Sprintf("Exists: message=%v", u.messageID.ShortID())
+func (u *targetedExists) String() string {
+	return fmt.Sprintf("TargetedExists: message = %v remote = %v targetStateID = %v", u.resp.messageID.InternalID.ShortID(), u.resp.messageID.RemoteID, u.targetStateID)
+}
+
+// ExistsStateUpdate needs to be a separate update since it has to deal with a Recent flag propagation. If a session
+// with a selected state appends a message, only that state should see the recent flag. If a message is appended to a
+// non-selected mailbox or arrives from remote, the first state with the selected mailbox should get the flag.
+// See ExistsStateUpdate.Apply() for more info.
+type ExistsStateUpdate struct {
+	lock sync.Mutex
+	MBoxIDStateFilter
+	responders    []*exists
+	targetStateID int
+}
+
+func NewExistsStateUpdate(mailboxID imap.InternalMailboxID, messageIDs []ids.MessageIDPair, uids map[imap.InternalMessageID]*ent.UID, s *State) Update {
+	var stateID int
+
+	if s != nil {
+		stateID = s.StateID
+	}
+
+	responders := xslices.Map(messageIDs, func(messageID ids.MessageIDPair) *exists {
+		uid := uids[messageID.InternalID]
+		exists := newExists(ids.NewMessageIDPair(uid.Edges.Message), uid.UID, db.NewFlagSet(uid, uid.Edges.Message.Edges.Flags))
+
+		return exists
+	})
+
+	return &ExistsStateUpdate{
+		MBoxIDStateFilter: MBoxIDStateFilter{MboxID: mailboxID},
+		responders:        responders,
+		targetStateID:     stateID,
+	}
+}
+
+func newExistsStateUpdateWithExists(mailboxID imap.InternalMailboxID, responders []*exists, s *State) Update {
+	var stateID int
```

Comment:
risky, can this potentially match the first state ID if we pass in a nil state? because i think the first state ID that the backend generates is `0`, which would be the default value here.

Reply:
We increment the stateID before using it. But you are right, could become an issue with overflow.

Reply:
ah okay so it starts at 0 but the first state gets 1. Might be worth using a pointer so we avoid using in-band values here.

Reply:
I added a bool safeguard, safer than keeping around pointers which may no longer exist.